sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/aio/_client.py

```diff
@@ -0,0 +1,1091 @@
+# coding=utf-8
+# --------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------
+# pylint: disable=client-method-missing-tracing-decorator-async,too-many-lines,networking-import-outside-azure-core-transport
+import asyncio
+import inspect
+from typing import Any, overload, Callable, Union, Optional, Dict, List, Literal, Awaitable
+import time
+import logging
+import urllib.parse
+import aiohttp
+from azure.core.pipeline.policies import RetryMode
+
+from ..models._models import (
+    OnConnectedArgs,
+    OnDisconnectedArgs,
+    OnServerDataMessageArgs,
+    OnGroupDataMessageArgs,
+    WebPubSubJsonProtocol,
+    WebPubSubJsonReliableProtocol,
+    SequenceId,
+    RetryPolicy,
+    WebPubSubGroup,
+    SendMessageErrorOptionsAsync,
+    WebPubSubMessage,
+    SendMessageError,
+    SendEventMessage,
+    SendToGroupMessage,
+    AckMessage,
+    ConnectedMessage,
+    SequenceAckMessage,
+    CloseEvent,
+    OnRejoinGroupFailedArgs,
+    DisconnectedMessage,
+    GroupDataMessage,
+    ServerDataMessage,
+    JoinGroupMessage,
+    LeaveGroupMessage,
+    AckMessageError,
+    AckMapAsync,
+    OpenClientError,
+)
+from ..models._enums import (
+    WebPubSubDataType,
+    WebPubSubClientState,
+    CallbackType,
+    WebPubSubProtocolType,
+)
+from .._client import (
+    WebPubSubClientCredential,
+    _RETRY_TOTAL,
+    _RETRY_BACKOFF_FACTOR,
+    _RETRY_BACKOFF_MAX,
+    _RECOVERY_TIMEOUT,
+    _RECOVERY_RETRY_INTERVAL,
+    _USER_AGENT,
+    _ACK_TIMEOUT,
+    _START_TIMEOUT,
+)
+from .._util import format_user_agent, raise_for_empty_message_ack, NO_ACK_MESSAGE_ERROR
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class WebSocketAppAsync:
+    def __init__(
+        self, url, on_open=None, subprotocols: Optional[List[str]] = None, header: Optional[Dict[str, str]] = None
+    ) -> None:
+        self.url = url
+        self.on_open = on_open
+        self.subprotocols = subprotocols
+        self.header = header
+        self.session: Optional[aiohttp.ClientSession] = None
+        self.sock: Optional[aiohttp.client._WSRequestContextManager] = None
+        self.event: asyncio.Event = asyncio.Event()
+        self.close_timeout = 10.0
+
+    async def receive(self) -> aiohttp.WSMessage:
+        return await self.sock.receive()  # type: ignore
+
+    async def send(self, message: str) -> None:
+        await self.sock.send_str(message)  # type: ignore
+
+    async def run_forever(self):
+        await self.on_open()
+
+    async def connect(self):
+        self.session = aiohttp.ClientSession()
+        self.sock = await self.session.ws_connect(self.url, protocols=self.subprotocols, headers=self.header)
```

Comment:
This should be wrapped in a `try ... except ClientConnectorError` in case a connection can't be made.

Reply:
Updated